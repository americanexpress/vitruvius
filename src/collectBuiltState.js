/*
 * Copyright (c) 2017 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

export default function collectBuiltState({ reducers, locals, defaultState }) {
  let builtState = defaultState;

  Object.keys(reducers).forEach((key) => {
    const reducer = reducers[key];
    if (typeof reducer === 'function' && typeof reducer.buildInitialState === 'function') {
      if (typeof builtState.set === 'function') {
        builtState = builtState.set(key, reducer.buildInitialState(locals));
      } else {
        builtState[key] = reducer.buildInitialState(locals);
      }
    }
  });

  return builtState;
}
