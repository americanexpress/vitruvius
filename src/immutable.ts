/*
 * Copyright (c) 2017 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

import { combineReducers } from 'redux-immutable';
import { Map } from 'immutable';
import collectBuiltState, { VitruviusImmutableReducers } from './collectBuiltState';

// internal type for use before assigning buildInitial state
type VitruviusImmutableOptionalCombinedReducer<Locals> = ReturnType<typeof combineReducers> & {
  buildInitialState?: (_locals: Locals) => Map<string, unknown>;
};

export type VitruviusImmutableCombinedReducer<Locals> = ReturnType<typeof combineReducers> & {
  buildInitialState: (_locals: Locals) => Map<string, unknown>;
};

export default function vitruviusImmutable<Locals>(
  reducers: VitruviusImmutableReducers<Locals>,
  getDefaultState = Map<string, unknown>
): VitruviusImmutableCombinedReducer<Locals> {
  const combined: VitruviusImmutableOptionalCombinedReducer<Locals> = combineReducers(
    reducers,
    getDefaultState
  ) as VitruviusImmutableOptionalCombinedReducer<Locals>;

  combined.buildInitialState = function buildInitialState(locals): Map<string, unknown> {
    return collectBuiltState<Locals, VitruviusImmutableReducers<Locals>>({
      reducers,
      locals,
      defaultState: getDefaultState(),
    });
  };

  return combined as VitruviusImmutableCombinedReducer<Locals>;
}
