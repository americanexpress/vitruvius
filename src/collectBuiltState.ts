/*
 * Copyright (c) 2017 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

import { Reducer } from 'redux';

import { Map } from 'immutable';

/* eslint-disable-next-line @typescript-eslint/no-explicit-any -- Here, we see a good use of any.

usually `unknown` should be used in the case that the caller hands in something of unknown type.

However, in this case, it's not the something the caller gave us to being broadened to unknown,
it (would be) the caller narrowing unknown to something concrete. This would lead to typescript
throwing an error.

Therefore, any must be used, so that when the caller tries to hand us a reducer, typescript does not say
'This reducer should be <unknown, unknown, unknown> but its <something, something, something>, that's not allowed'.
But instead says 'aah, this should be <any, any, any>, and <something, something, something> is a valid subtype of that'.
 */
export type VitruviusReducer<Locals> = Reducer<any, any, any> & {
  buildInitialState?: (_locals: Locals) => Record<string, unknown>;
};

export type VitruviusReducers<Locals> = Record<string, VitruviusReducer<Locals>>;

// eslint-disable-next-line @typescript-eslint/no-explicit-any -- see above
export type VitruviusImmutableReducer<Locals> = Reducer<any, any, any> & {
  buildInitialState?: (_locals: Locals) => Map<string, unknown>;
};

export type VitruviusImmutableReducers<Locals> = Record<string, VitruviusImmutableReducer<Locals>>;

export default function collectBuiltState<
  Locals,
  TR extends VitruviusReducers<Locals> | VitruviusImmutableReducers<Locals>
>({
  reducers,
  locals,
  defaultState,
}: {
  reducers: TR extends VitruviusImmutableReducers<Locals>
    ? VitruviusImmutableReducers<Locals>
    : VitruviusReducers<Locals>;
  locals: Locals;
  defaultState: TR extends VitruviusImmutableReducers<Locals>
    ? Map<string, unknown>
    : Record<string, unknown>;
}): TR extends VitruviusImmutableReducers<Locals> ? Map<string, unknown> : Record<string, unknown> {
  let builtState = defaultState;

  Object.keys(reducers).forEach((key) => {
    const reducer = reducers[key];
    if (typeof reducer === 'function' && typeof reducer.buildInitialState === 'function') {
      if (typeof builtState.set === 'function') {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment -- This is the case where the users is using immutableJS However, to narrow the type here to 'typeof buildState === immutable.Map' would be a breaking change, since currently this code allows for any object with a 'set' method.
        builtState = builtState.set(key, reducer.buildInitialState(locals));
      } else {
        // @ts-expect-error -- since we cant strictly narrow the if clause above to be the Map case, we also cant strictly narrow the else clause here to be the Record case
        builtState[key] = reducer.buildInitialState(locals);
      }
    }
  });

  return builtState;
}
