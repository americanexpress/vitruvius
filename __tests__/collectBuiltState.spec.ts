/*
 * Copyright (c) 2017 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

import { describe, it, expect } from 'vitest';
import { Map } from 'immutable';
import collectBuiltState from '../src/collectBuiltState';
import { stuffReducer, thingsReducer, otherReducer, staticReducer } from './fixtures';

describe('collectBuiltState', () => {
  const locals = { data: 'test' };

  it('should build an immutable state', () => {
    const result = collectBuiltState({
      reducers: {
        stuff: stuffReducer,
        things: thingsReducer,
        other: otherReducer,
        static: staticReducer,
      },
      locals,
      defaultState: Map(),
    });
    expect(result).toMatchSnapshot();
  });

  it('should build a mutable state', () => {
    const result = collectBuiltState({
      reducers: {
        stuff: stuffReducer,
        things: thingsReducer,
        other: otherReducer,
        static: staticReducer,
      },
      locals,
      defaultState: {},
    });
    expect(result).toMatchSnapshot();
  });
});
