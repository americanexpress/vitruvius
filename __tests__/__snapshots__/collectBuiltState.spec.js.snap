// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collectBuiltState should build a mutable state 1`] = `
Object {
  "other": Immutable.Map {
    baz: "test",
},
  "stuff": Immutable.Map {
    foo: "test",
},
  "things": Immutable.Map {
    bar: "test",
},
}
`;

exports[`collectBuiltState should build an immutable state 1`] = `
Immutable.Map {
  stuff: Immutable.Map {
    foo: "test",
  },
  things: Immutable.Map {
    bar: "test",
  },
  other: Immutable.Map {
    baz: "test",
  },
}
`;
